### PR TITLE
Fix Permissioning In DatabricksPySparkStepLauncher

### PR DESCRIPTION
**Pull request by matt-weingarten**
This fixes the permission gaps in DatabricksPySparkStepLauncher:

1. Using the updated Databricks API
2. Not missing attaching to the cluster entirely while it's being spun up

## Summary & Motivation

This is related to https://github.com/dagster-io/dagster/issues/25056. This fixes the permissioning gap that exists with the DatabricksPySparkStepLauncher, so that permissions are actually set correctly for non-admin users.

## How I Tested These Changes

We tested this locally by overriding the Dagster method with these changes, and permissions are now being set properly.

## Changelog

Updating the step launcher logic with updated Databricks API calls.

- [ ] `NEW` _(added new feature or capability)_
- [x] `BUGFIX` _(fixed a bug)_
- [ ] `DOCS` _(added or updated documentation)_

### CODE REVIEW

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py
Comment (smats0n) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:429]: it can take several moments to initialize a run and the cluster_instance so this check running immediately after the run submission cause the run to fail almost always

Screenshot 2024-10-04 at 10 42 54 AM

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py
Comment (smats0n) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:430]: since Databricks Jobs v2.1 API, the `cluster_instance` is set on the individual task of a job run. The `run_info.cluster_instance` is always returned as None. I believe it exists for backward compatibility. Since the dagster_databricks client is using the new v2.1 format, we need to access `cluster_info` for the run from the `tasks` object.

Screenshot 2024-10-04 at 10 41 35 AM
Screenshot 2024-10-04 at 10 40 41 AM

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py
Comment (smats0n) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:448]: @matt-weingarten can you retain this check?

File: python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py
Comment (smats0n) [inline python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:450]: @matt-weingarten can you retain this casting?

**Review by smats0n — COMMENTED** [review]
looks good to me 👍

**Comment by smats0n** [comment]
@garethbrickman, I noticed a review request for me. These looks good to me but Im only a contributor to the project. I think we'll need a review from a project maintainer :)

**Comment by yuhan** [comment]
@danielgafni do you mind reviewing this PR?

**Review by danielgafni — APPROVED** [review]
Looking good to me, perhaps @smackesey could run the local tests?

**Comment by smackesey** [comment]
The local tests are only for pipes, which doesn't use the step launcher.

BK however is giving a type error:

```
  | [2024-10-14T15:52:44Z] /workdir/python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:
  | [2024-10-14T15:52:44Z]   430:29: Object of type "None" is not subscriptable (reportOptionalSubscript)
  | [2024-10-14T15:52:44Z]
  | [2024-10-14T15:52:44Z]   430:64: "cluster_id" is not a known attribute of "None" (reportOptionalMemberAccess)
  | [2024-10-14T15:52:44Z]
```

@matt-weingarten if you fix that error and repush we can merge

**Comment by matt-weingarten** [comment]
> The local tests are only for pipes, which doesn't use the step launcher.
> 
> BK however is giving a type error:
> 
> ```
>   | [2024-10-14T15:52:44Z] /workdir/python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py:
>   | [2024-10-14T15:52:44Z]   430:29: Object of type "None" is not subscriptable (reportOptionalSubscript)
>   | [2024-10-14T15:52:44Z]
>   | [2024-10-14T15:52:44Z]   430:64: "cluster_id" is not a known attribute of "None" (reportOptionalMemberAccess)
>   | [2024-10-14T15:52:44Z]
> ```
> 
> @matt-weingarten if you fix that error and repush we can merge

Could this because of a version difference in Databricks? Not sure how to diagnose the difference between this and what there was before (since that's apparently working).

**Comment by danielgafni** [comment]
Might be a typing quirk of databricks. I suggest running `pyright` against a few `databricks` versions to check for this. 

But you should probably just do something like this:

```python
if run_info.tasks is not None and run_info.tasks[0].cluster_instance is not None:
    # your code
else:
    raise DagsterInvariantViolationError(f"run_info {run_info} doesn't contain cluster_id")
```

**Comment by matt-weingarten** [comment]
> Might be a typing quirk of databricks. I suggest running `pyright python_modules/libraries/dagster-databricks` against a few `databricks` versions to check for this.
> 
> But you should probably just do something like this:
> 
> ```python
> if run_info.tasks is not None and run_info.tasks[0].cluster_instance is not None:
>     # your code
> else:
>     raise DagsterInvariantViolationError(f"run_info {run_info} doesn't contain cluster_id")
> ```

Added!

**Comment by smackesey** [comment]
@matt-weingarten please run `make ruff` from repo root and repush, that should get BK to green

**Comment by matt-weingarten** [comment]
> @matt-weingarten please run `make ruff` from repo root and repush, that should get BK to green

Done!

**Pull request merged**
